File: apps/android/core/protocol/src/test/kotlin/com/wrapdrive/core/protocol/ConformanceTest.kt
package com.wrapdrive.core.protocol

import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import java.io.File
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * Cross-language conformance against the shared test vectors in
 * `protocol-spec/test-vectors/`.
 *
 * For every fixture this asserts the Kotlin serializer emits exactly the
 * vector's `canonicalJson` and that parsing that JSON reproduces the canonical
 * message. The TypeScript suite runs the same vectors, so the two
 * implementations stay byte-compatible.
 *
 * Validates: Requirements 7.3, 7.5, 8.2
 */
class ConformanceTest : StringSpec({
    val vectorsDir = locateVectorsDir()
    val json = Json { ignoreUnknownKeys = true }

    val files = vectorsDir.listFiles { f -> f.extension == "json" }?.sortedBy { it.name } ?: emptyList()

    "test vectors are present" {
        files.isNotEmpty() shouldBe true
    }

    files.forEach { file ->
        val vector = json.parseToJsonElement(file.readText()).jsonObject
        val type = vector["type"]!!.jsonPrimitive.content
        val canonical = vector["canonicalJson"]!!.jsonPrimitive.content
        val message = vector["message"]!!.jsonObject

        "${file.name}: serializes message to canonical JSON" {
            reserializeByType(type, message.toString()) shouldBe canonical
        }

        "${file.name}: parses canonical JSON and re-serializes identically" {
            reserializeByType(type, canonical) shouldBe canonical
        }
    }
})

/** Parse JSON of the given type into its model, then serialize canonically. */
private fun reserializeByType(type: String, jsonText: String): String =
    when (type) {
        "DeviceInfo" -> ProtocolJson.serialize(ProtocolJson.parseDeviceInfo(jsonText))
        "Capabilities" -> ProtocolJson.serialize(ProtocolJson.parseCapabilities(jsonText))
        "TransferPlan" -> ProtocolJson.serialize(ProtocolJson.parseTransferPlan(jsonText))
        "FileMeta" -> ProtocolJson.serialize(ProtocolJson.parseFileMeta(jsonText))
        "ChunkRef" -> ProtocolJson.serialize(ProtocolJson.parseChunkRef(jsonText))
        "PrepareUploadRequest" ->
            ProtocolJson.serialize(ProtocolJson.parsePrepareUploadRequest(jsonText))
        "PrepareUploadResult" ->
            ProtocolJson.serialize(ProtocolJson.parsePrepareUploadResult(jsonText))
        "TransferProgress" -> ProtocolJson.serialize(ProtocolJson.parseTransferProgress(jsonText))
        else -> error("unknown vector type $type")
    }

/** Walk up from the working directory to find protocol-spec/test-vectors. */
private fun locateVectorsDir(): File {
    val start = System.getProperty("user.dir") ?: "."
    var dir: File? = File(start)
    while (dir != null) {
        val candidate = File(dir, "protocol-spec/test-vectors")
        if (candidate.isDirectory) return candidate
        dir = dir.parentFile
    }
    error("could not locate protocol-spec/test-vectors from ${System.getProperty("user.dir")}")
}

File: packages/protocol-ts/src/conformance.test.ts
/**
 * Protocol conformance against the shared test vectors.
 *
 * For every fixture in `protocol-spec/test-vectors/`, this asserts that the
 * TypeScript serializer emits exactly the vector's `canonicalJson` and that
 * parsing that JSON reproduces the canonical `message`. The Kotlin suite runs
 * the same vectors, so the two implementations stay byte-compatible.
 *
 * Validates: Requirements 7.3, 7.5, 8.2
 */

import { readFileSync, readdirSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, join, resolve } from 'node:path';
import { describe, expect, it } from 'vitest';
import {
  parseCapabilities,
  parseChunkRef,
  parseDeviceInfo,
  parseFileMeta,
  parsePrepareUploadRequest,
  parsePrepareUploadResult,
  parseTransferPlan,
  parseTransferProgress,
} from './serialization.js';
import {
  serializeCapabilities,
  serializeChunkRef,
  serializeDeviceInfo,
  serializeFileMeta,
  serializePrepareUploadRequest,
  serializePrepareUploadResult,
  serializeTransferPlan,
  serializeTransferProgress,
} from './serialization.js';

interface Vector {
  type: string;
  description: string;
  message: unknown;
  canonicalJson: string;
}

const serializers: Record<string, (msg: never) => string> = {
  DeviceInfo: serializeDeviceInfo as (msg: never) => string,
  Capabilities: serializeCapabilities as (msg: never) => string,
  TransferPlan: serializeTransferPlan as (msg: never) => string,
  FileMeta: serializeFileMeta as (msg: never) => string,
  ChunkRef: serializeChunkRef as (msg: never) => string,
  PrepareUploadRequest: serializePrepareUploadRequest as (msg: never) => string,
  PrepareUploadResult: serializePrepareUploadResult as (msg: never) => string,
  TransferProgress: serializeTransferProgress as (msg: never) => string,
};

const parsers: Record<string, (json: string) => unknown> = {
  DeviceInfo: parseDeviceInfo,
  Capabilities: parseCapabilities,
  TransferPlan: parseTransferPlan,
  FileMeta: parseFileMeta,
  ChunkRef: parseChunkRef,
  PrepareUploadRequest: parsePrepareUploadRequest,
  PrepareUploadResult: parsePrepareUploadResult,
  TransferProgress: parseTransferProgress,
};

const vectorsDir = resolve(
  dirname(fileURLToPath(import.meta.url)),
  '..',
  '..',
  '..',
  'protocol-spec',
  'test-vectors',
);

function loadVectors(): Array<{ file: string; vector: Vector }> {
  return readdirSync(vectorsDir)
    .filter((f) => f.endsWith('.json'))
    .map((file) => ({
      file,
      vector: JSON.parse(readFileSync(join(vectorsDir, file), 'utf8')) as Vector,
    }));
}

describe('protocol conformance against shared test vectors', () => {
  const vectors = loadVectors();

  it('found vector fixtures', () => {
    expect(vectors.length).toBeGreaterThan(0);
  });

  for (const { file, vector } of vectors) {
    describe(`${file} (${vector.type})`, () => {
      it('serializes to the canonical JSON byte-for-byte', () => {
        const serialize = serializers[vector.type];
        expect(serialize, `no serializer for type ${vector.type}`).toBeDefined();
        expect(serialize(vector.message as never)).toBe(vector.canonicalJson);
      });

      it('parses the canonical JSON back to the message', () => {
        const parse = parsers[vector.type];
        expect(parse, `no parser for type ${vector.type}`).toBeDefined();
        expect(parse(vector.canonicalJson)).toEqual(vector.message);
      });
    });
  }
});
